AsciiString optimized hashCode

Scottmitch opened the pull request:
Motivation:
The AsciiString.hashCode() method can be optimized. This method is frequently used while to build the DefaultHeaders data structure.

Modification:
- Add a PlatformDependent hashCode algorithm which utilizes UNSAFE if available

Result:
AsciiString hashCode is faster.

Scottmitch commented:
Here are the benchmark results:

```
Benchmark                                          (size)   Mode  Cnt          Score         Error  Units
AsciiStringBenchmark.hashCodeBenchBytesNew              3  thrpt   10  171620686.411 ± 2067177.087  ops/s
AsciiStringBenchmark.hashCodeBenchBytesNew              5  thrpt   10  175861229.576 ± 3144324.972  ops/s
AsciiStringBenchmark.hashCodeBenchBytesNew              8  thrpt   10  160477033.757 ± 3303639.514  ops/s
AsciiStringBenchmark.hashCodeBenchBytesNew             10  thrpt   10  158626444.748 ± 1986456.633  ops/s
AsciiStringBenchmark.hashCodeBenchBytesNew             20  thrpt   10  127607141.124 ± 2167646.826  ops/s
AsciiStringBenchmark.hashCodeBenchBytesNew             50  thrpt   10   72125526.060 ± 2106200.452  ops/s
AsciiStringBenchmark.hashCodeBenchBytesOld              3  thrpt   10  176451138.408 ± 2384382.162  ops/s
AsciiStringBenchmark.hashCodeBenchBytesOld              5  thrpt   10  143466864.042 ± 1759612.341  ops/s
AsciiStringBenchmark.hashCodeBenchBytesOld              8  thrpt   10  109661110.010 ± 1154123.205  ops/s
AsciiStringBenchmark.hashCodeBenchBytesOld             10  thrpt   10  102299314.252 ± 1008274.904  ops/s
AsciiStringBenchmark.hashCodeBenchBytesOld             20  thrpt   10   62262391.840 ± 1098801.523  ops/s
AsciiStringBenchmark.hashCodeBenchBytesOld             50  thrpt   10   26230133.517 ±  357407.087  ops/s
AsciiStringBenchmark.hashCodeBenchCharSequenceNew       3  thrpt   10  183354375.923 ± 3443497.773  ops/s
AsciiStringBenchmark.hashCodeBenchCharSequenceNew       5  thrpt   10  158780695.388 ± 1622008.061  ops/s
AsciiStringBenchmark.hashCodeBenchCharSequenceNew       8  thrpt   10  107521051.275 ± 3779937.365  ops/s
AsciiStringBenchmark.hashCodeBenchCharSequenceNew      10  thrpt   10  100615689.343 ± 1529300.849  ops/s
AsciiStringBenchmark.hashCodeBenchCharSequenceNew      20  thrpt   10   72518474.366 ± 1715818.963  ops/s
AsciiStringBenchmark.hashCodeBenchCharSequenceNew      50  thrpt   10   38357354.866 ±  440889.222  ops/s
AsciiStringBenchmark.hashCodeBenchCharSequenceOld       3  thrpt   10  166837329.528 ± 5377966.980  ops/s
AsciiStringBenchmark.hashCodeBenchCharSequenceOld       5  thrpt   10  132701651.724 ± 2552045.916  ops/s
AsciiStringBenchmark.hashCodeBenchCharSequenceOld       8  thrpt   10  101663377.948 ± 1089496.581  ops/s
AsciiStringBenchmark.hashCodeBenchCharSequenceOld      10  thrpt   10   96768184.705 ± 1655143.710  ops/s
AsciiStringBenchmark.hashCodeBenchCharSequenceOld      20  thrpt   10   59591234.529 ±  653538.861  ops/s
AsciiStringBenchmark.hashCodeBenchCharSequenceOld      50  thrpt   10   25853228.738 ±  384579.786  ops/s
```

Scottmitch commented:
Here is the result of a `wrk` benchmark:

before optimized hash code:
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    15.91ms   22.58ms 874.22ms   85.96%
    Req/Sec   197.29k    30.08k  496.01k    78.03%
  374716668 requests in 2.00m, 50.60GB read
Requests/sec: 3120069.60
Transfer/sec:    431.45MB
```

with this PR:
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    15.00ms   21.30ms 846.37ms   86.23%
    Req/Sec   210.93k    35.51k  355.74k    71.98%
  400726492 requests in 2.00m, 54.11GB read
Requests/sec: 3336640.04
Transfer/sec:    461.40MB
```

Scottmitch commented:
In terms of how evenly the algorithm distributes I ran a collision test with the new and old algorithms.  I grabbed a list of 58120 English ascii words offline, plus netty's HTTP and HTTP/2 pseudo headers names, and then some headers found in the wild.

Collisions with old hash code
```
19007732 : excellency tract
```

Collisions with new hash code
```
777808595 : eagles moles 
808232867 : duvets sweets 
804164183 : makers users 
-840554786 : haulers hitters 
1566404641 : moated united 
2082282164 : efface menace 
-1292817698 : hauling hitting 
-1150087593 : making using 
257675040 : laager tiger 
```

The hash code collision rate is higher with the new algorithm for this data set. However non of the header values collided so It may not matter (feedback/comments welcome).

Scottmitch commented on codec-http/src/test/java/io/netty/handler/codec/http/HashCollisionDetector.java:
```diff
@@ -0,0 +1,144 @@
+import java.util.Map.Entry;
+import java.util.Set;
+
+public class HashCollisionDetector {
```
The plan is to not submit this class.  It is just used for hash code collision detection.

normanmaurer replied:
ensure to remove before merge.

Scottmitch replied:
yip removing now.

Scottmitch commented on common/src/main/java/io/netty/util/internal/PlatformDependent.java:
```diff
@@ -361,6 +366,69 @@ public static long getLong(byte[] data, int index) {
+    }
+
+    private static short getShortFromBytesSafe(CharSequence bytes, int offset) {
+        return (short) (bytes.charAt(offset) & 0xFF | (bytes.charAt(offset + 1) << 8));
```
I need to think about endianess here...

normanmaurer replied:
yeah I think we need to handle endianness

Scottmitch commented:
@normanmaurer @buchgr - FYI

normanmaurer commented on common/src/main/java/io/netty/util/internal/PlatformDependent.java:
```diff
@@ -361,6 +366,69 @@ public static long getLong(byte[] data, int index) {
+                    ((long) bytes[offset + 6] & 0xff) <<  8 |
+                    (long) bytes[offset + 7] & 0xff;
+        }
+        return (long) bytes[offset] & 0xFF |
```
just a nit but you use `0xff` everywhere else so please use it here as well.

Scottmitch replied:
sgtm. I will use all lowercase for hex numbers.

normanmaurer commented on common/src/main/java/io/netty/util/internal/PlatformDependent.java:
```diff
@@ -361,6 +366,69 @@ public static long getLong(byte[] data, int index) {
+    }
+
+    private static long getLongFromBytesSafe(CharSequence bytes, int offset) {
+        return (long) bytes.charAt(offset) & 0xFF |
```
same as above

normanmaurer replied:
is it correct that we not care about endianness here ?

Scottmitch replied:
https://github.com/netty/netty/pull/4436#discussion_r43830826 applies to all of the `get*FromBytesSafe` methods. So yes I need to think about endianess for all of them :)

normanmaurer commented on common/src/main/java/io/netty/util/internal/PlatformDependent.java:
```diff
@@ -361,6 +366,69 @@ public static long getLong(byte[] data, int index) {
+                    (bytes[offset + 2] & 0xff) <<  8 |
+                    bytes[offset + 3] & 0xff;
+        }
+        return bytes[offset] & 0xFF |
```
same as above

normanmaurer commented on common/src/main/java/io/netty/util/internal/PlatformDependent.java:
```diff
@@ -361,6 +366,69 @@ public static long getLong(byte[] data, int index) {
+    }
+
+    private static int getIntFromBytesSafe(CharSequence bytes, int offset) {
+        return bytes.charAt(offset) & 0xFF |
```
same as above

normanmaurer commented on common/src/main/java/io/netty/util/internal/PlatformDependent.java:
```diff
@@ -361,6 +366,69 @@ public static long getLong(byte[] data, int index) {
+
+    private static short getShortSafe(byte[] bytes, int offset) {
+        if (BIG_ENDIAN_NATIVE_ORDER) {
+            return (short) (bytes[offset] << 8 | (bytes[offset + 1] & 0xFF));
```
same as above

normanmaurer commented on common/src/main/java/io/netty/util/internal/PlatformDependent.java:
```diff
@@ -361,6 +366,69 @@ public static long getLong(byte[] data, int index) {
+        if (BIG_ENDIAN_NATIVE_ORDER) {
+            return (short) (bytes[offset] << 8 | (bytes[offset + 1] & 0xFF));
+        }
+        return (short) (bytes[offset] & 0xFF | (bytes[offset + 1] << 8));
```
same as above

normanmaurer commented on common/src/main/java/io/netty/util/internal/PlatformDependent0.java:
```diff
@@ -272,35 +299,186 @@ static void copyMemory(Object src, long srcOffset, Object dst, long dstOffset, l
+        // code the same regardless of character case (upper case or lower case hash is the same).
+        return (hash * 31 +
+                // High order int
+                (int) ((value & 0x1F1F1F1F00000000L) >>> 32)) * 31 +
```
nice trick ;)

Scottmitch replied:
:). Getting rid of `ByteString` and having makes it simpler to just use a case insensitive hash code for `AsciiString`.

normanmaurer commented:
@Scottmitch just a few comments. Overall LGTM.

Scottmitch commented on common/src/main/java/io/netty/util/internal/PlatformDependent.java:
```diff
@@ -361,6 +366,78 @@ public static long getLong(byte[] data, int index) {
+
+    private static long getLongFromBytesSafe(CharSequence bytes, int offset) {
+        if (BIG_ENDIAN_NATIVE_ORDER) {
+            // TODO
```
I need to fix these before shipping...

normanmaurer replied:
go slacker ;)

Scottmitch replied:
I know....be cool to be able to test it. Anyone willing to donate a big endian server to the build slave pool?

normanmaurer replied:
yeah I agree :(

Scottmitch commented on common/src/main/java/io/netty/util/internal/PlatformDependent.java:
```diff
@@ -361,6 +366,91 @@ public static long getLong(byte[] data, int index) {
+
+    private static long getLongFromBytesSafe(CharSequence bytes, int offset) {
+        if (BIG_ENDIAN_NATIVE_ORDER) {
+            // Swap the bytes within each char, and don't invert the way the array is indexed.
```
@normanmaurer - Can you double check the byte swapping for the 3 `get*FromBytesSafe` methods here (they all take `CharSequence` parameters)?

Scottmitch commented:
I tweaked the algorithm a bit and updated the performance results. The collisions were due to not dispersing the "least significant bytes" of the `byte[]` enough. I added additional left rotate operations when dealing with the last remaining input to get better dispersion. Now there are no collisions in this test set.

normanmaurer commented on common/src/main/java/io/netty/util/internal/PlatformDependent.java:
```diff
@@ -361,6 +366,91 @@ public static long getLong(byte[] data, int index) {
+    private static long getLongFromBytesSafe(CharSequence bytes, int offset) {
+        if (BIG_ENDIAN_NATIVE_ORDER) {
+            // Swap the bytes within each char, and don't invert the way the array is indexed.
+            return (long) ((bytes.charAt(offset) & 0xff00) >>> 8) |
```
why you do the ">>> 8" here ?

Scottmitch replied:
this was removed.

Scottmitch commented:
I verified this PR on little endian unsafe/safe and also big endian safe. The big endian was emulated  (qemu, debian 7, ppc 32bit) and didn't provide unsafe, but this may be good enough to move forward.  All code that is accounting for endianness explicitly is verified correct, there may just be some code in the unsafe implementation that is not accounting for endiannes when it should be.

@normanmaurer - I think we have been operating under the assumption "review the endinaness code, and if it is broken someone will let us know". This PR takes it a step further and partially verifies correctness...is this good enough to merge?

normanmaurer commented:
@Scottmitch fine with me.

normanmaurer commented on common/src/main/java/io/netty/util/internal/PlatformDependent0.java:
```diff
@@ -272,35 +299,188 @@ static void copyMemory(Object src, long srcOffset, Object dst, long dstOffset, l
+            return UNSAFE.getByte(bytes1, baseOffset1) == UNSAFE.getByte(bytes2, baseOffset2);
+        default:
+            return true;
+        }
```
I think it would be better to have a `case 0:` that returns true and throw an error other wise.

Scottmitch replied:
I thought about this, but could really justify it to myself. We control the value of `remainingBytes` in this method's scope, when could it be anything else? It seems like we would just be adding a "dead" code path.

normanmaurer replied:
ok fine... just leave it as it is and cherry-pick

Scottmitch commented:
Cherry-picked 4.1 (b4b7913) master (33b74e3)
